tests/conftest.py

```diff
@@ -136,3 +137,20 @@ def pytest_generate_tests(metafunc):
 
         elif "accelerators" in metafunc.fixturenames:
             metafunc.parametrize("accelerators", ACCELERATORS)
+
+
+@pytest.fixture
+def numba_threads():
+    from numba import get_num_threads, set_num_threads
+
+    original_threads = get_num_threads()
+
+    @contextmanager
+    def nthreads(n: int):
+        set_num_threads(n)
+        try:
+            yield
+        finally:
+            set_num_threads(original_threads)
+
+    yield nthreads
```

Comment:
Since `get_threads()` takes no input even on the other side, you may simplify this to
```suggestion
    set_num_threads(get_threads())
    yield
    set_num_threads(original_threads)
```

To be fair, the code above is even doing something more than preventing the context manager in the unit test. The context manager you used is resetting the original number of threads in case an error happens, and only in that case. However, I would always reset the state. And, with fixtures, you do not need to explicitly handle exceptions, since the Pytest machinery will capture the error anyhow, and always execute the finalization block (i.e. resume the execution after the `yield`).

Reply:
Moreover, this hack was strictly related to Quimb, and I have no evidence that the current numba fixture may be useful for anything else beyond that.

So, you could even use something like:
```py
from numba import get_num_threads, set_num_threads

@pytest.fixture
def quimb_backend():
    original_threads = get_num_threads()
    set_num_threads(get_threads())
    from qibotn.backends.quimb import QuimbBackend
    yield QuimbBackend
    set_num_threads(original_threads)
```
and put it directly in the `test_backends_qibotn.py` file.
Fixtures are scoped, you can make them available for all tests, but it is not worth to expose them in the most global scope, if they are needed only in a very specific place.

Btw, to the best of my knowledge is the Quimb import to be troublesome, not the Numba one - so that one can stay top-level.

Reply:
> The context manager you used is resetting the original number of threads in case an error happens

shouldn't the ``finally`` always happen regardless of the ``try`` outcome (which anyway should never fail since is just yielding none)? 
In general I liked more the idea to call ``with numba_threads(n):`` as a context manager. it conveys more the idea of some scoped resources setting, rather than plainly calling  ``numba_threads(n)``. Furthermore, the resetting of the threads is happening only at the end of the test in the second case, which is less modular (not necessarily worse, just a choice).

In any case, having the fixture to return the ``QuimbBackend`` works as well, but I don't find it very useful since it's used only there. I mean, at that point we could just manually set the threads in that specific test without involving a fixture.

Reply:
> shouldn't the `finally` always happen regardless of the `try` outcome (which anyway should never fail since is just yielding none)?

Whoops, you're right: I saw a `try:` and inferred an `except:`. My bad.

Reply:
> In general I liked more the idea to call `with numba_threads(n):` as a context manager. it conveys more the idea of some scoped resources setting, rather than plainly calling `numba_threads(n)`. Furthermore, the resetting of the threads is happening only at the end of the test in the second case, which is less modular (not necessarily worse, just a choice).

Well, but then you put the whole test in the context manager...
In the end, you're creating an extra scope, which is identical to the one you already had. And the `fixture` is already doing the job of the context manager, handling the finalization.

So, you may argue to keep the context manager instead of the fixture. Which is fair - I just find the fixture easier/more suitable for a test, but they will perform the exact same task.

> In any case, having the fixture to return the `QuimbBackend` works as well, but I don't find it very useful since it's used only there.

Well, it is used only there in any case.

> I mean, at that point we could just manually set the threads in that specific test without involving a fixture.

This is just an argument against modularity: they would be performing two different tasks, so it is relevant to split into different functions.
I do not consider a great practice having long functions, just because the other ones you may write would be used in a single place...

Reply:
> Well, but then you put the whole test in the context manager...

yes, in this case, but nothing prevents you from having a tests with different sections using a different number of threads, which is not possible with the other approach.

> Well, it is used only there in any case.

True, but more likely to be reused in other places in the future, rather than the ``quimb_backend``, at least this is my feeling.

Reply:
> True, but more likely to be reused in other places in the future, rather than the `quimb_backend`, at least this is my feeling.

Ok, I agree that is more likely to be reused than the other. And we may just have a different perception.

My point is that it is just unlikely, in absolute terms, that even the Numba one will be reused.
The need for setting the number of threads in an unscoped manner, outside a backend or anything else, is not something that I expect we're going to test for.

In particular, we would not be using that even here, it was just a workaround to avoid Quimb failing (if you remove that, it seemed to fail - but it looked like a Quimb/Autoray bug, that may even be gone by now...).

Reply:
So, the test used to fail here, in qibo, and that is the reason why we were manually setting the threads, right? Because I tried right now without any manual threads setting and it passes.

Reply:
I think so - either that, or it was failing in Qibotn, and then it got copy-pasted here.

However, I believe it is possible it was failing even before the test, directly during the import.

If it's not failing (or not failing any longer), better not doing anything, and keep all the imports top-level. Mostly as you did in 445013e (but you can lift the `QuimbBackend` import as well).